### PR TITLE
multi: bump neutrino+btcwallet versions, use new generic lru cache

### PR DESCRIPTION
**Pull request by Roasbeef**
In this PR, we bump the btcwallet+neutrino versions to allow the project to use the new generic lru cache that was added to the `neutrino` package. With this, we'll get additional type safety as we don't need to assert/cast back and forth when we want to retrive items from the cache. 

This is the first part to resolving: https://github.com/lightningnetwork/lnd/issues/7396

One issue I encountered while trying to fix that above is that: we want to key the cache by height, and then map to a slice of items. We then append items each time w message for that height appears, and we'll iterate the entire thing to reprocess once we catch up to that height. 

The main issue here is that if we want to add 130 items, but the cache capacity is 100 items, then this always fails. One fix would be to flatten the cache, and store just the individual messages. The issue with this is that the current logic assumes the height as a key, since that lets us iterate over the cache easily. We could go with the scid here, but then we'd need to iterate the number of transactions in a block (and output index, etc) if we want to retain the same logic.

### CODE REVIEW

File: discovery/gossiper.go [hunk 1]
```diff
@@ -1863,7 +1869,7 @@ func (d *AuthenticatedGossiper) isPremature(chanID lnwire.ShortChannelID,
 	result, err := d.futureMsgs.Get(msgHeight)
 	// No error returned means we have old messages cached.
 	if err == nil {
-		cachedMsgs = result.(*cachedNetworkMsg)
+		cachedMsgs = result
```
Comment (yyforyongyu) [inline discovery/gossiper.go:1872]: should we fix the heap escape here too?
``` diff
diff --git a/discovery/gossiper.go b/discovery/gossiper.go
index fd3c8ca43..ae7ca83f0 100644
--- a/discovery/gossiper.go
+++ b/discovery/gossiper.go
@@ -1888,7 +1888,10 @@ func (d *AuthenticatedGossiper) isPremature(chanID lnwire.ShortChannelID,
 
        // Add the network message.
        cachedMsgs.msgs = append(cachedMsgs.msgs, pMsg)
-       _, err = d.futureMsgs.Put(msgHeight, cachedMsgs)
+       _, err = d.futureMsgs.Put(
+               msgHeight, &cachedNetworkMsg{
+                       msgs: cachedMsgs.msgs,
+               })
        if err != nil {
                log.Errorf("Adding future message got error: %v", err)
        }
```

File: discovery/gossiper.go [hunk 2]
```diff
@@ -469,17 +469,23 @@ type AuthenticatedGossiper struct {
...
+		selfKey:     selfKeyDesc.PubKey,
+		selfKeyLoc:  selfKeyDesc.KeyLocator,
+		cfg:         &cfg,
+		networkMsgs: make(chan *networkMsg),
+		futureMsgs: lru.NewCache[uint32, *cachedNetworkMsg](
+			maxPrematureUpdates,
+		),
+		quit:              make(chan struct{}),
+		chanPolicyUpdates: make(chan *chanPolicyUpdateRequest),
+		prematureChannelUpdates: lru.NewCache[uint64, *cachedNetworkMsg](
```
Comment (guggero) [inline discovery/gossiper.go:481]: needs a `//notlint:lll` line above to silence the linter.

**Review by guggero — APPROVED** [review]
LGTM pending the linter fix and @Crypt-iQ's comment.

File: discovery/gossiper.go (hunk 1 above)
Comment (Crypt-iQ) [reply discovery/gossiper.go:1872]: yes

**Comment by Roasbeef** [comment]
So I went to implement the suggestion of: 
  * Maintain a map of height to cache key 
  * Use that to iterate over to find out which items in the cache are now ready 

I ran into an issue where you actually need to keep those two maps (the map of keys in the cache keyed by height, and the cache itself) in sync. Eg: if you add something new to the cache, rn it doesn't tell you which item was excited, so then you'd need to iterate again to see what's actually in the cache. 

I think what we need here is a richer set of methods on the cache, similar to `sync.Map`. We want to be able to iterate over the cache, and also get/delete in the same step. With that, I think we'd be able to implement the above.

**Review by yyforyongyu — APPROVED** [review]
LGTM🙏

File: discovery/gossiper.go (hunk 1 above)
Comment (Crypt-iQ) [reply discovery/gossiper.go:1872]: needs:
```
diff --git a/discovery/gossiper.go b/discovery/gossiper.go
index e8183472e..0c021ddda 100644
--- a/discovery/gossiper.go
+++ b/discovery/gossiper.go
@@ -1887,9 +1887,10 @@ func (d *AuthenticatedGossiper) isPremature(chanID lnwire.ShortChannelID,
        pMsg := &processedNetworkMsg{msg: copied}
 
        // Add the network message.
-       cachedMsgs.msgs = append(cachedMsgs.msgs, pMsg)
+       msgs := cachedMsgs.msgs
+       msgs = append(msgs, pMsg)
        _, err = d.futureMsgs.Put(msgHeight, &cachedNetworkMsg{
-               msgs: cachedMsgs.msgs,
+               msgs: msgs,
        })
        if err != nil {
                log.Errorf("Adding future message got error: %v", err)
```

**Pull request merged**